From 0fe5a234240a108b294a87174ad197f6b5cb38e9 Mon Sep 17 00:00:00 2001
From: Paul Stewart <pstew@google.com>
Date: Thu, 3 Mar 2016 15:40:19 -0800
Subject: [PATCH 3/5] Remove newlines from wpa_supplicant config network
 output

Spurious newlines output while writing the config file can corrupt the
wpa_supplicant configuration. Avoid writing these for the network block
parameters. This is a generic filter that cover cases that may not have
been explicitly addressed with a more specific commit to avoid control
characters in the psk parameter.

Signed-off-by: Paul Stewart <pstew@google.com>

Modified for OpenXT
Signed-off-by: Christopher Clark <christopher.clark6@baesystems.com>
---

diff --git a/src/utils/common.c b/src/utils/common.c
index a28dd5b..ebb8336 100644
--- a/src/utils/common.c
+++ b/src/utils/common.c
@@ -372,3 +372,13 @@ int has_ctrl_char(const u8 *data, size_t len)
    }
    return 0;
 }
+
+int has_newline(const char *str)
+{
+    while (*str) {
+        if (*str == '\n' || *str == '\r')
+            return 1;
+        str++;
+    }
+    return 0;
+}
diff --git a/src/utils/common.h b/src/utils/common.h
index 18e77ab..ce58cb5 100644
--- a/src/utils/common.h
+++ b/src/utils/common.h
@@ -444,6 +444,7 @@ int wpa_snprintf_hex(char *buf, size_t buf_size, const u8 *data, size_t len);
 int wpa_snprintf_hex_uppercase(char *buf, size_t buf_size, const u8 *data,
 			       size_t len);
 int has_ctrl_char(const u8 *data, size_t len);
+int has_newline(const char *str);
 
 #ifdef CONFIG_NATIVE_WINDOWS
 void wpa_unicode2ascii_inplace(TCHAR *str);
diff --git a/wpa_supplicant/config.c b/wpa_supplicant/config.c
index 4ec0f26..8508ecf 100644
--- a/wpa_supplicant/config.c
+++ b/wpa_supplicant/config.c
@@ -1963,8 +1963,19 @@ char * wpa_config_get(struct wpa_ssid *ssid, const char *var)
 
 	for (i = 0; i < NUM_SSID_FIELDS; i++) {
 		const struct parse_data *field = &ssid_fields[i];
-		if (os_strcmp(var, field->name) == 0)
-			return field->writer(field, ssid);
+		if (os_strcmp(var, field->name) == 0) {
+			char *ret = field->writer(field, ssid);
+
+			if (ret && has_newline(ret)) {
+				wpa_printf(MSG_ERROR,
+						"Found newline in value for %s; not returning it",
+						var);
+				os_free(ret);
+				ret = NULL;
+			}
+
+			return ret;
+		}
 	}
 
 	return NULL;
